security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/AbstractClientCredentialsClient.java

```diff
@@ -0,0 +1,127 @@
+
+    /**
+     * @param tokenEndpointClient The token endpoint client
+     * @param oauthClientConfiguration The client configuration
+     */
+    public AbstractClientCredentialsClient(OauthClientConfiguration oauthClientConfiguration,
+                                          TokenEndpointClient tokenEndpointClient) {
+        this.oauthClientConfiguration = oauthClientConfiguration;
+        this.tokenEndpointClient = tokenEndpointClient;
+    }
+
+    @NonNull
+    @Override
+    public Publisher<TokenResponse> clientCredentials(@Nullable String scope) {
+        return clientCredentials(scope, false);
+    }
+
+    @Override
+    @NonNull
+    public Publisher<TokenResponse> clientCredentials(@Nullable String scope, boolean force) {
+        String resolvedScope = scope != null ? scope : NOSCOPE;
+        if (!scopeToPublisherMap.containsKey(resolvedScope)) {
+            scopeToPublisherMap.put(resolvedScope, new CacheablePublisher<>(TokenResponseExpiration::new));
+        }
+        CacheablePublisher<TokenResponse> publisher = scopeToPublisherMap.get(resolvedScope);
+        if (force || isExpired(publisher.getElement())) {
+            publisher.clear();
+        }
+        if (publisher.getSubscription() == null) {
+            ClientCredentialsTokenRequestContext context = createTokenRequestContext(scope);
+            tokenEndpointClient.sendRequest(context).subscribe(publisher);
+        }
+        return publisher;
+    }
+
+    /**
+     *
+     * @param tokenResponse Token Response
+     * @return true if a parameter token response is null or if an expiration time cannot parsed. false if the expiration time is after the current date.
+     */
+    protected boolean isExpired(@Nullable TokenResponse tokenResponse) {
+        if (tokenResponse == null) {
+            return true;
+        }
+        return expirationDate(tokenResponse).map(expTime -> {
+            final Date now = new Date();
+            return expTime.before(now);
+        }).orElse(true);
+    }
+
+    /**
+     *
+     * @param tokenResponse Token Response
+     * @return The expiration date from the exp claim in the access token is a JWT or the expiration date calculated from the expiresIn
+     */
+    protected Optional<Date> expirationDate(@NonNull TokenResponse tokenResponse) {
+        try {
+            JWT jwt = JWTParser.parse(tokenResponse.getAccessToken());
+            return Optional.ofNullable(jwt.getJWTClaimsSet().getExpirationTime());
+        } catch (ParseException e) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("cannot parse access token {} to JWT", tokenResponse.getAccessToken());
+            }
+        }
+        if (tokenResponse instanceof TokenResponseExpiration) {
+            TokenResponseExpiration tokenResponseExpiration = (TokenResponseExpiration) tokenResponse;
+            return Optional.ofNullable(tokenResponseExpiration.getExpiration());
+        }
```

Comment:
I think the Json Web Token expiration date claim is more precise than the `expiresIn` value. 

The `expiresIn` is the time to live in seconds for the token since the authorization server issued it. However, there is a network latency involved. I cannot really know when the token was issued. Thus, the expiration date calculated with the `expiresIn` is more blurry. 

Exp claim is an epoch value. Thus, I think it is slightly more precise.